sycl/test/extensions/annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,243 @@
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note %s
+// expected-no-diagnostics
+
+// Common tests for annotated USM allocation
+
+#include "sycl/sycl.hpp"
+
+#include <iostream>
+
+// clang-format on
+
+using namespace sycl::ext::oneapi::experimental;
+using namespace sycl::ext::intel::experimental;
+using alloc = sycl::usm::alloc;
+
+constexpr int N = 10;
+
+void TestAlloc(sycl::queue &q) {
+  // P1: input properties
+  properties P1{conduit, stable, cache_config{large_slm}};
+  // ANNP1: output properties with usm_kind and no runtime property
+  // `cache_config`
+  properties ANNP1{conduit, stable, usm_kind<alloc::device>};
+  auto APtr1 = malloc_device_annotated<int>(N, q, P1);
+  auto APtr11 = malloc_device_annotated<int, decltype(P1)>(N, q, P1);
+  auto APtr12 =
+      malloc_device_annotated<int, decltype(P1), decltype(ANNP1)>(N, q, P1);
+  // Check that the annotated_ptr allocated should contain properties in ANNP1
+  static_assert(
+      std::is_same_v<decltype(APtr1), annotated_ptr<int, decltype(ANNP1)>>);
+  static_assert(std::is_same_v<decltype(APtr11), decltype(APtr11)>);
+  static_assert(std::is_same_v<decltype(APtr12), decltype(APtr12)>);
+
+  // P2: input properties
+  properties P2{conduit, buffer_location<5>};
+  // ANNP2: output properties with usm_kind
+  properties ANNP2{conduit, buffer_location<5>, usm_kind<alloc::device>};
+  auto APtr2 = malloc_device_annotated<int>(N, q, P2);
+  static_assert(
+      std::is_same_v<decltype(APtr2), annotated_ptr<int, decltype(ANNP2)>>);
+
+  // P3: input properties
+  properties P3{conduit, stable, cache_config{large_data}};
+  // ANNP3: output properties with usm_kind and no runtime property
+  // `cache_config`
+  properties ANNP3{conduit, stable, usm_kind<alloc::device>};
+  auto APtr3 = malloc_device_annotated<int>(N, q, P3);
+  static_assert(
+      std::is_same_v<decltype(APtr3), annotated_ptr<int, decltype(ANNP3)>>);
+
+  // APtr4 is of type annotated_ptr<void, decltype(properties{conduit, stable,
+  // usm_kind<sycl::usm::alloc::device>})>
+  auto APtr4 = malloc_device_annotated(N, q, P2);
+  auto APtr41 = malloc_device_annotated<decltype(P2)>(N, q);
+  auto APtr42 = malloc_device_annotated<decltype(P2), decltype(ANNP2)>(N, q);
+  static_assert(
+      std::is_same_v<decltype(APtr4), annotated_ptr<void, decltype(ANNP2)>>);
+  static_assert(std::is_same_v<decltype(APtr4), decltype(APtr41)>);
+  static_assert(std::is_same_v<decltype(APtr4), decltype(APtr42)>);
+
+  // APtr3 and APtr4 are not same because the underlying pointer type is int* vs
+  // void*
```

Comment:
Another reason APtr3 and APtr4 are not the same, is that they have different properties. Maybe fix this test case so the only thing different is the datatype T.